source/MaterialXGenShader/ColorManagementSystem.cpp

```diff
@@ -0,0 +1,92 @@
+#include <MaterialXGenShader/ColorManagementSystem.h>
+
+#include <MaterialXGenShader/ShaderGenerator.h>
+#include <MaterialXGenShader/Nodes/SourceCodeNode.h>
+
+namespace MaterialX
+{
+
+ColorSpaceTransform::ColorSpaceTransform(const string& ss, const string& ts, const TypeDesc& t)
```

Comment:
The way TypeDesc is implemented it's ment to be used only for a set of unique instances (one per registered type) and passed as raw pointers to these instances. In this case I see your point in passing by reference since the type is required, but for TypeDesc I think it's more clear and consistent if we always use raw pointers. And in this case you are throwing anyway if the type is not equal to COLOR3 or COLOR4, so a null pointer check is already done implicitly here.

Reply:
If I'm following, you want me to change the signature to take a TypeDesc pointer instead of a reference and to change the code to:
```
if (!type)
{
    throw ExceptionShaderGenError("Color space transform can only be a color3 or color4, nullptr requested.");
}
else if (*type != Type::COLOR3 && *type != Type::COLOR4)
{
    throw ExceptionShaderGenError("Color space transform can only be a color3 or color4, '" + type->getName() + "' requested.");
}
```

Is that right?

Reply:
No simpler:
Type::COLOR3/4 are pointers so the only thing you need to change is the signature, from a TypeDesc ref to a TypeDesc pointer.
If null is passed in it will then fall into the same guard, being an invalid type.

Reply:
I don't think that will quite work though as I'm throwing the name of the type that was passed which will dereference nullptr.

Reply:
No sure you need to include the invalid type name in the message. Only stating that the type is not a color3 or color4 should be enough I think.

Reply:
Ok